src/main/java/org/apache/sysml/hops/ipa/IPAPassFlagDimensionPreserveFunctionOp.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sysml.hops.ipa;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.Map.Entry;
+
+import org.apache.sysml.conf.ConfigurationManager;
+import org.apache.sysml.hops.FunctionOp;
+import org.apache.sysml.hops.HopsException;
+import org.apache.sysml.lops.FunctionCallCPSingle;
+import org.apache.sysml.parser.DMLProgram;
+import org.apache.sysml.parser.FunctionStatementBlock;
+import org.apache.sysml.parser.LanguageException;
+import org.apache.sysml.parser.StatementBlock;
+
+/**
+ * A rewrite to split DAGs after FunctionOps that return matrices/frames
+ * and are not dimension-preserving.
+ *
+ */
+public class IPAPassFlagDimensionPreserveFunctionOp extends IPAPass {
```

Comment:
Again, let's not focus on IPA right now, but there is already `isUnarySizePreservingFunction` in `InterProceduralAnalysis` - so it would be good to reuse that.